File: task-7444be/environment/gmpb_native/gmpb.c
/*
 * Generalized Moving Peaks Benchmark (GMPB) — C implementation.
 *
 * Based on: D. Yazdani et al., "Benchmarking continuous dynamic optimization:
 * Survey and generalized test suite," IEEE Trans. Cybernetics, 2020.
 *
 */

#include "gmpb.h"
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <stdint.h>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

/* ------------------------------------------------------------------ */
/*  xorshift64* PRNG — simple, fast, deterministic                    */
/* ------------------------------------------------------------------ */
typedef struct { uint64_t s; } rng_t;

static void rng_init(rng_t *r, uint64_t seed) {
    r->s = seed ? seed : 1ULL;
    for (int i = 0; i < 20; i++) {          /* warm-up */
        r->s ^= r->s >> 12;
        r->s ^= r->s << 25;
        r->s ^= r->s >> 27;
    }
}

static uint64_t rng_next(rng_t *r) {
    uint64_t x = r->s;
    x ^= x >> 12;
    x ^= x << 25;
    x ^= x >> 27;
    r->s = x;
    return x * UINT64_C(0x2545F4914F6CDD1D);
}

static double rng_uniform(rng_t *r) {
    return (rng_next(r) >> 11) / 9007199254740992.0;   /* [0, 1) */
}

static double rng_uniform_range(rng_t *r, double lo, double hi) {
    return lo + (hi - lo) * rng_uniform(r);
}

/* Box-Muller transform */
static double rng_normal(rng_t *r) {
    double u1, u2;
    do { u1 = rng_uniform(r); } while (u1 < 1e-300);
    u2 = rng_uniform(r);
    return sqrt(-2.0 * log(u1)) * cos(2.0 * M_PI * u2);
}

/* ------------------------------------------------------------------ */
/*  Helper math                                                       */
/* ------------------------------------------------------------------ */
static double vec_norm(const double *v, int n) {
    double s = 0.0;
    for (int i = 0; i < n; i++) s += v[i] * v[i];
    return sqrt(s);
}

static double dclamp(double x, double lo, double hi) {
    return x < lo ? lo : (x > hi ? hi : x);
}

/* ------------------------------------------------------------------ */
/*  GMPB internal structure                                           */
/* ------------------------------------------------------------------ */
struct gmpb_t {
    int    num_peaks;
    int    dimension;
    int    change_frequency;
    double shift_severity;
    int    num_environments;
    double height_severity;   /* 7.0 */
    double width_severity;    /* 1.0 */
    double min_coord;         /* -25.0 */
    double max_coord;         /*  25.0 */

    int    total_evals;
    int    eval_count;
    int    current_env;
    double best_found;

    double *positions;        /* num_peaks * dimension, row-major */
    double *heights;          /* num_peaks */
    double *widths;           /* num_peaks */
    double *prev_shifts;      /* num_peaks * dimension */

    double error_sum;         /* running sum for offline-error mean */
    rng_t  rng;
};

/* ------------------------------------------------------------------ */
/*  Environment change                                                */
/* ------------------------------------------------------------------ */
static void change_environment(gmpb_t *g) {
    g->current_env++;
    g->best_found = -1e300;

    int dim = g->dimension;
    int np  = g->num_peaks;

    double *rand_vec = (double *)malloc(dim * sizeof(double));
    double *combined = (double *)malloc(dim * sizeof(double));
    if (!rand_vec || !combined) { free(rand_vec); free(combined); return; }

    for (int i = 0; i < np; i++) {
        /* Random unit direction */
        for (int d = 0; d < dim; d++)
            rand_vec[d] = rng_normal(&g->rng);
        double rn = vec_norm(rand_vec, dim);
        if (rn > 1e-15)
            for (int d = 0; d < dim; d++) rand_vec[d] /= rn;

        /* Combine with previous shift (correlation) */
        double *prev = g->prev_shifts + i * dim;
        for (int d = 0; d < dim; d++)
            combined[d] = prev[d] + rand_vec[d];

        double cn = vec_norm(combined, dim);
        if (cn > 1e-15) {
            for (int d = 0; d < dim; d++)
                prev[d] = combined[d] / cn * g->shift_severity;
        } else {
            for (int d = 0; d < dim; d++)
                prev[d] = rand_vec[d] * g->shift_severity;
        }

        /* Update position with boundary clamping */
        double *pos = g->positions + i * dim;
        for (int d = 0; d < dim; d++) {
            pos[d] += prev[d];
            pos[d] = dclamp(pos[d], g->min_coord, g->max_coord);
        }

        /* Perturb height within [10, 100] */
        g->heights[i] += g->height_severity * rng_normal(&g->rng);
        g->heights[i] = dclamp(g->heights[i], 10.0, 100.0);

        /* Perturb width within [0.5, 15] */
        g->widths[i] += g->width_severity * rng_normal(&g->rng);
        g->widths[i] = dclamp(g->widths[i], 0.5, 15.0);
    }

    free(rand_vec);
    free(combined);
}

/* ------------------------------------------------------------------ */
/*  Public API                                                        */
/* ------------------------------------------------------------------ */
gmpb_t *gmpb_new(int num_peaks, int dimension, int change_frequency,
                  double shift_severity, int num_environments,
                  unsigned int seed) {
    gmpb_t *g = (gmpb_t *)calloc(1, sizeof(gmpb_t));
    if (!g) return NULL;

    g->num_peaks        = num_peaks;
    g->dimension        = dimension;
    g->change_frequency = change_frequency;
    g->shift_severity   = shift_severity;
    g->num_environments = num_environments;
    g->height_severity  = 7.0;
    g->width_severity   = 1.0;
    g->min_coord        = -25.0;
    g->max_coord        =  25.0;

    g->total_evals = change_frequency * num_environments;
    g->eval_count  = 0;
    g->current_env = 0;
    g->best_found  = -1e300;
    g->error_sum   = 0.0;

    rng_init(&g->rng, (uint64_t)seed);

    int np  = num_peaks;
    int dim = dimension;

    g->positions   = (double *)malloc(np * dim * sizeof(double));
    g->heights     = (double *)malloc(np * sizeof(double));
    g->widths      = (double *)malloc(np * sizeof(double));
    g->prev_shifts = (double *)calloc(np * dim, sizeof(double));

    if (!g->positions || !g->heights || !g->widths || !g->prev_shifts) {
        gmpb_free(g);
        return NULL;
    }

    /* Initialize peak parameters */
    for (int i = 0; i < np; i++) {
        for (int d = 0; d < dim; d++)
            g->positions[i * dim + d] =
                rng_uniform_range(&g->rng, g->min_coord, g->max_coord);
        g->heights[i] = rng_uniform_range(&g->rng, 30.0, 70.0);
        g->widths[i]  = rng_uniform_range(&g->rng, 1.0, 12.0);
    }

    return g;
}

void gmpb_free(gmpb_t *g) {
    if (!g) return;
    free(g->positions);
    free(g->heights);
    free(g->widths);
    free(g->prev_shifts);
    free(g);
}

double gmpb_evaluate(gmpb_t *g, const double *x) {
    if (!g || !x) return NAN;
    if (g->eval_count >= g->total_evals) return NAN;

    int dim = g->dimension;
    int np  = g->num_peaks;

    /* f(x) = max_i { h_i - w_i * ||x - c_i|| } */
    double best_fit = -1e300;
    for (int i = 0; i < np; i++) {
        double dist2 = 0.0;
        const double *ci = g->positions + i * dim;
        for (int d = 0; d < dim; d++) {
            double diff = x[d] - ci[d];
            dist2 += diff * diff;
        }
        double val = g->heights[i] - g->widths[i] * sqrt(dist2);
        if (val > best_fit) best_fit = val;
    }

    /* Update best-found */
    if (best_fit > g->best_found) g->best_found = best_fit;

    /* Current optimum = max peak height */
    double optimum = -1e300;
    for (int i = 0; i < np; i++)
        if (g->heights[i] > optimum) optimum = g->heights[i];

    /* Record current error */
    double error = optimum - g->best_found;
    if (error < 0.0) error = 0.0;
    g->error_sum += error;

    g->eval_count++;

    /* Trigger environment change at boundary */
    if (g->eval_count % g->change_frequency == 0 &&
        g->eval_count < g->total_evals) {
        change_environment(g);
    }

    return best_fit;
}

int gmpb_is_finished(const gmpb_t *g) {
    return g ? (g->eval_count >= g->total_evals) : 1;
}

int gmpb_get_environment(const gmpb_t *g) {
    return g ? g->current_env : -1;
}

double gmpb_get_offline_error(const gmpb_t *g) {
    if (!g || g->eval_count == 0) return INFINITY;
    return g->error_sum / g->eval_count;
}

int gmpb_get_eval_count(const gmpb_t *g) {
    return g ? g->eval_count : 0;
}

int gmpb_get_remaining_evals(const gmpb_t *g) {
    return g ? (g->total_evals - g->eval_count) : 0;
}

int gmpb_get_dimension(const gmpb_t *g) {
    return g ? g->dimension : 0;
}

int gmpb_get_num_peaks(const gmpb_t *g) {
    return g ? g->num_peaks : 0;
}

int gmpb_get_change_frequency(const gmpb_t *g) {
    return g ? g->change_frequency : 0;
}

double gmpb_get_min_coord(const gmpb_t *g) {
    return g ? g->min_coord : 0.0;
}

double gmpb_get_max_coord(const gmpb_t *g) {
    return g ? g->max_coord : 0.0;
}

double gmpb_get_optimum(const gmpb_t *g) {
    if (!g) return NAN;
    double opt = -1e300;
    for (int i = 0; i < g->num_peaks; i++)
        if (g->heights[i] > opt) opt = g->heights[i];
    return opt;
}

int gmpb_get_optimum_position(const gmpb_t *g, double *pos_out) {
    if (!g || !pos_out) return -1;
    int best_idx = 0;
    for (int i = 1; i < g->num_peaks; i++)
        if (g->heights[i] > g->heights[best_idx]) best_idx = i;
    memcpy(pos_out, g->positions + best_idx * g->dimension,
           g->dimension * sizeof(double));
    return 0;
}

File: task-7444be/environment/gmpb.py
"""
Generalized Moving Peaks Benchmark (GMPB) for Dynamic Optimization.

Based on: D. Yazdani et al., "Benchmarking continuous dynamic optimization:
Survey and generalized test suite," IEEE Transactions on Cybernetics, 2020.

Generates dynamic fitness landscapes composed of cone-shaped peaks with
controllable characteristics. The landscape changes periodically, requiring
optimizers to track the moving optimum across environment transitions.

The fitness function is:
    f(x) = max_{i=1..m} { h_i - w_i * ||x - c_i|| }

where h_i, w_i, c_i are the height, width, and center of peak i.

After every `change_frequency` evaluations, peak positions shift by
`shift_severity` in a correlated random direction, and heights/widths
undergo bounded random perturbations.
"""

import numpy as np


class GMPB:
    """Generalized Moving Peaks Benchmark.

    Attributes:
        num_peaks (int): Number of peaks in the landscape.
        dimension (int): Dimensionality of the search space.
        change_frequency (int): Evaluations between environment changes.
        shift_severity (float): Magnitude of peak position shifts per change.
        num_environments (int): Total number of environments.
        min_coord (float): Lower bound of search space (-25.0).
        max_coord (float): Upper bound of search space (25.0).
        height_severity (float): Std dev of height perturbation per change.
        width_severity (float): Std dev of width perturbation per change.
    """

    def __init__(self, num_peaks=10, dimension=5, change_frequency=5000,
                 shift_severity=1.0, num_environments=30, seed=42,
                 height_severity=7.0, width_severity=1.0):
        self.num_peaks = num_peaks
        self.dimension = dimension
        self.change_frequency = change_frequency
        self.shift_severity = shift_severity
        self.num_environments = num_environments
        self.height_severity = height_severity
        self.width_severity = width_severity
        self.min_coord = -25.0
        self.max_coord = 25.0

        self._rng = np.random.RandomState(seed)
        self._total_evals = change_frequency * num_environments
        self._eval_count = 0
        self._current_env = 0
        self._current_errors = []
        self._best_found = -np.inf

        # Initialize peak parameters
        self._positions = self._rng.uniform(
            self.min_coord, self.max_coord, (num_peaks, dimension))
        self._heights = self._rng.uniform(30.0, 70.0, num_peaks)
        self._widths = self._rng.uniform(1.0, 12.0, num_peaks)
        self._prev_shifts = np.zeros((num_peaks, dimension))

    @property
    def optimum(self):
        """Current global optimum fitness value."""
        return float(np.max(self._heights))

    @property
    def optimum_position(self):
        """Position of the current global optimum peak."""
        return self._positions[np.argmax(self._heights)].copy()

    def evaluate(self, x):
        """Evaluate the fitness of a candidate solution.

        Computes the maximum over all cone-shaped peak contributions,
        updates the best-found tracking, records the current error,
        and triggers an environment change if the evaluation count
        reaches a multiple of change_frequency.

        Args:
            x: Array-like of shape (dimension,) within [min_coord, max_coord].

        Returns:
            float: The fitness value at x.

        Raises:
            RuntimeError: If the evaluation budget is exhausted.
            ValueError: If x has the wrong dimensionality.
        """
        if self._eval_count >= self._total_evals:
            raise RuntimeError("Evaluation budget exhausted")

        x = np.asarray(x, dtype=np.float64).ravel()
        if x.shape[0] != self.dimension:
            raise ValueError(
                f"Expected {self.dimension}-d vector, got {x.shape[0]}-d")

        # Vectorized fitness: f(x) = max_i { h_i - w_i * ||x - c_i|| }
        diffs = self._positions - x[np.newaxis, :]
        dists = np.sqrt(np.sum(diffs ** 2, axis=1))
        peaks = self._heights - self._widths * dists
        fitness = float(np.max(peaks))

        # Update best-found and record current error
        self._best_found = max(self._best_found, fitness)
        error = max(0.0, self.optimum - self._best_found)
        self._current_errors.append(error)
        self._eval_count += 1

        # Trigger environment change at boundary
        if (self._eval_count % self.change_frequency == 0 and
                self._eval_count < self._total_evals):
            self._change_environment()

        return fitness

    def _change_environment(self):
        """Apply stochastic perturbations to create a new environment."""
        self._current_env += 1
        self._best_found = -np.inf

        for i in range(self.num_peaks):
            # Correlated random shift direction
            rand_vec = self._rng.randn(self.dimension)
            rand_vec /= (np.linalg.norm(rand_vec) + 1e-15)
            combined = self._prev_shifts[i] + rand_vec
            norm = np.linalg.norm(combined)
            if norm > 1e-15:
                shift = combined / norm * self.shift_severity
            else:
                shift = rand_vec * self.shift_severity
            self._prev_shifts[i] = shift

            # Update position with boundary clamping
            self._positions[i] += shift
            self._positions[i] = np.clip(
                self._positions[i], self.min_coord, self.max_coord)

            # Perturb height within [10, 100]
            self._heights[i] += self.height_severity * self._rng.randn()
            self._heights[i] = np.clip(self._heights[i], 10.0, 100.0)

            # Perturb width within [0.5, 15]
            self._widths[i] += self.width_severity * self._rng.randn()
            self._widths[i] = np.clip(self._widths[i], 0.5, 15.0)

    def get_offline_error(self):
        """Compute offline error: mean current error over all evaluations.

        Returns:
            float: The offline error, or inf if no evaluations performed.
        """
        if not self._current_errors:
            return float('inf')
        return float(np.mean(self._current_errors))

    def is_finished(self):
        """Check if the evaluation budget is exhausted.

        Returns:
            bool: True if eval_count >= change_frequency * num_environments.
        """
        return self._eval_count >= self._total_evals

    def get_environment(self):
        """Get the current environment index (0-based).

        Returns:
            int: Current environment number.
        """
        return self._current_env

    def get_eval_count(self):
        """Get the total number of evaluations performed so far.

        Returns:
            int: Evaluation count.
        """
        return self._eval_count

    def get_remaining_evals(self):
        """Get the remaining evaluation budget.

        Returns:
            int: Remaining evaluations.
        """
        return self._total_evals - self._eval_count
